### ISSUE TITLE
Pipelines

### ISSUE THREAD

**Issue by ArturilloD** [puggan/mc-factory#6]
Is there any chance of becoming a tileentity?

**Comment by puggan** [comment]
It is a title-entity already:
https://github.com/puggan/mc-factory/blob/0ac8b46d6e588b72d57f654335842204844557fe/src/main/java/se/puggan/factory/container/FactoryEntity.java#L34

**Comment by puggan** [comment]
Is there a specific mod you want me to test it with, searched for "Pipelines" but didn't found one.

**Comment by ArturilloD** [comment]
Sorry to wait for the reply and for the brevity of the previous message.
The mod I was referring to is the Pretty Pipes (
https://www.curseforge.com/minecraft/mc-mods/pretty-pipes), I'm playing
version 1.15.2 of Minecraft, 31.2.36 of Forge , 1.5.4 from Pretty Pipes and
1.1.1 from Workbench Factory.

I have been doing some creative testing with some other mods installed, I
was trying to make a kind of factory to automate large amounts of items,
with the pipes I can distribute better in a certain way, since the Pretty
Pipes mod is not exact to the time to distribute items, since it is
distributing as they disappear in the inventory that gives it demand. The
point is that its mod is incredible, I like it very much, it is the best
automatic manufacturing table I have found so far, the AutoCrafting one is
very good, since it serves as a work table regardless of whether it is
automatic, it has its own inventory and others, although the bad thing is
that it requires redstone to work and also generates a lot of lag, another
of the bad things is that so far all the automatic work tables that I have
found always have a loose item out there and cannot be recovered all items
unless manually removed.

In the tests that I have been doing, apart from what I have already
mentioned is that when admitting a stack of items the first slots are
always filled and since they are not distributed equally, stacks of lost
items remain, I suppose a better solution would be that it could only enter
one item per slot, although there will still be items lying around, since
it does not require all the items to enter, that is, the complete recipe.
Another thing that I have detected is that the Refined Relocation 2 mod (
https://www.curseforge.com/minecraft/mc-mods/refined-relocation-2) has a
hopper that is created with redstone and allows stacks to be moved whole,
the hopper cannot put items on the table as with the pipes mod, so I
suppose it is due to the same problem and it is that it is not a
tileEntity, what it has sent me before I do not understand it, already I
unfortunately have no idea about Java programming, I will try to do some
tests.

I think I did not leave anything and excuse me for the paragraph and the
Indian, although that is Google's fault, heh. I wish I knew where to start
to be able to create things like that, I'm a little tired of having two
hundred mods for loose things and not being able to do my own. If I knew I
would help you, even so if you need something do not hesitate to contact
me. Greetings and thanks in advance.

**Comment by ArturilloD** [comment]
I've been trying for a while to decipher the code of the file that happened
to me in the previous email but that of the .jar that I have installed, but
... nothing at all, I just get Minecraft crashing. Who would say that I
program websites hehe

**Comment by puggan** [comment]
I'll play around with 2 mods you sugested, this weekend, and see if I can figure out whats happening.

As you seams to be intrested, Il try to decribing my findings in more details then usal :-)

**Comment by ArturilloD** [comment]
Ok, thank you very much, I'm going to fiddle with the code and see if I can
make at least some of the changes. Regards

**Comment by puggan** [comment]
Ah, my code that prevent hoppers to mess with the inventory, stops the pippeps from using the inventory.

Senario:
1. The hopper wants to send an item
2. Hopper checks if its pointing to a tile-entity, if not abort.
3. Hopper checks if its an inventory with CapabilityItemHandler.ITEM_HANDLER_CAPABILITY
4. if so, try to put the item in each slot in the inventory, done
5. if not, check if the inventory is a ISidedInventory
6. if so, ask what slots are avaible from the connected side
7. try to put the item in each avaible slot.

To prevent the hopper from puting/pulling item from the wrong slots, I answered that it dosn't have ITEM_HANDLER_CAPABILITY.

Pipes check for ITEM_HANDLER_CAPABILITY when it checks if it can connect.

**Comment by puggan** [comment]
AbstractFurnaceTileEntity using a SidedInvWrapper to solve the same problem,
but by answering that it has ITEM_HANDLER_CAPABILITY, and then act as a filter, 
if a tile entity have 3 slots (nr 0, 1 & 2), and only nr 2 is avaible from the bottom,
when accessing the tile entity throght SidedInvWrapper, getSlots() returns 1 insted of 3, and insertItem(0, ...) converts to insertItem(2, ...) and so on.

**Comment by puggan** [comment]
See row 177-182 in PipeBlock.java
https://github.com/Ellpeck/PrettyPipes/blob/294559ca614fdf2851d0277d6fbb130859f9684a/src/main/java/de/ellpeck/prettypipes/pipe/PipeBlock.java#L177-L182

And see row 334-339 in FactoryEntity.java
https://github.com/puggan/mc-factory/blob/0ac8b46d6e588b72d57f654335842204844557fe/src/main/java/se/puggan/factory/container/FactoryEntity.java#L334-L339

**Comment by puggan** [comment]
v1.2.0
![2020-08-26_15 59 20](https://user-images.githubusercontent.com/866668/91313006-2a1f1b80-e7b5-11ea-8fe3-6fa300d2f8a7.png)

**Comment by puggan** [comment]
v1.2.0 solves the connection issue,

It fills the stacks up in a odd way, the chosen receipt was a stone pickaxes, so it has 3 slots for stone.
When it started to fill it up I saw the following step of stones slots: [64, 0, 0] -> [64, 1, 0] -> [64, 64, 0] -> [64, 64, 1].
I'l see what I can do with that for version 1.2.1, but first I want to release the v1.2.0 version for all mc-versions i support.
But I keep this Issue open as it's not completely solved yet.

**Comment by ArturilloD** [comment]
It looks incredible, I've been looking at the lines and emails, they still
sound a bit like Chinese to me, I'll try to decipher all that to better
understand everything you want to explain.

So far what I understand is that in order to enter items and that the items
do not go with hoppers or pipes, it is necessary to establish that they can
be entered but only the X slot can remove items from the factory, is that
correct?

On the other hand, what he explained to me in the last post about inserting
items by stacks that he did before, that is, if he got to insert items in
the other slots but the previous one always had preference until it did not
have a stack per slot That's why yesterday I told him that I think it would
be interesting if he could only enter one item per slot, so that way I
think it would be more efficient when entering items with hoppers or pipes.
On the other hand, since it is not possible to extract items only from the
bottom of the Factory block, the problem that will continue to exist is
that if there are not enough items to be able to complete the recipe, items
will always remain inside the Factory, I do not know how complicated it can
be. be to fix that, although just in case I'm going to leave you a couple
of ideas here:
1.- It occurs to me that it could allow to extract items from the sides,
that way with pipes it would be easy to extract the items that have been
left over.
2.- The other is that since the Factory block "already knows" (I take it
for granted, since you have to specify in advance which recipe you want to
make), which recipe is required that does not allow you to enter items
until it detects that they are all necessary items in any of the adjacent
hoppers or chests.

The latter I understand that especially for pipes it must be quite complex
to carry out, I really hope I could help you, I would like that in the
future we could work together on a project.

Thanks for everything you really don't find many people like you anymore.

When can I try this version that you have shown me? So if you want I can
act as a pet for the tests you need to perform. Rest assured that your code
will be safe and you do not have to worry about that aspect. Regards

**Comment by puggan** [comment]
Oh, now I see 3 part of the problem:
1. The pipes didn't connect
2. An uneven fill, can make it produce less items, as there can be enough materials, but in the wrong slots.
3. For extracting items, it would be nice if the direction is ignored.

**Comment by ArturilloD** [comment]
Exactly, that's what I meant if the items do not fit well, you may expect a
stack of hay but you receive half, when you look at the table you see that
the first slot has a stack.
Sorry if you don't understand me very well, I'm using the translator.

**Comment by ArturilloD** [comment]
If you send me the code I can take some screenshots with what I want to
say, that way perhaps I can see more graphically what I am trying to say,
also you can get a better idea of where the problem may be.

**Comment by puggan** [comment]
The soulotion for "1. The pipes didn't connect"
was replace this code I found somewhare:
https://github.com/puggan/mc-factory/blob/0ac8b46d6e588b72d57f654335842204844557fe/src/main/java/se/puggan/factory/container/FactoryEntity.java#L334-L339

with the code at row 487-501 in AbstractFurnaceTileEntity.java
https://github.com/puggan/mc-factory/blob/9210a2661b65b75741682d58f53cec28ee9cf7e3/src/main/java/se/puggan/factory/container/FactoryEntity.java#L340-L353

**Comment by ArturilloD** [comment]
What did you find? Then I look at the code to compare what changes it has
made and see if I can understand something else. Right now I'm going to
test the changes that you've made, I'm dying to, heh.

**Comment by ArturilloD** [comment]
I have been doing some tests and well, it is incredible that pipes can
already be connected, the problem is that it continues to happen as you
said before that not all have the same priority, that is, they are filled
one by one, I suppose this could be solve in a certain way putting one item
per slot, the problem would still be there, since they would also be filled
in order, the point is that it would take much less time, since when
sending nine items at the same time from a hopper or a pipe it would fill
more Quick.
Great mod, congrats

**Comment by ArturilloD** [comment]
It just occurred to me that the easiest and safest solution, although safe
items will still be left over, although it can be controlled with pipes, is
that those nine spaces that are left when a recipe has already been
assigned that are from inventory, that is, that the order of the recipe is
not needed, but rather that there are simply enough items in those nine
slots to be able to generate the recipe

**Comment by puggan** [comment]
Had some time to test the current version, and some ides how to improve it, but not enough time to write the code yet.

On my mind right now:
* For an item that just use one type of material like 9 coal -> 1 coal block, the hoppers fill up the item-slots in a way that is efficient, but pipes send more then one at the time, so it fills them up in the same order, but with larger amount, therefor unbalanced.
* For an item that have more then one material like 1 stone + 2 stick -> 1 shovel, the hopper can't add sticks if there is no stone, I think that it checks the first slot, notice it's empty, ask if it can put the stick there, when the answer is no, it gives up.

* When trying to make the next item, If any of the required slot is empty, insted of aborting, I could add "balance material"-step, that spread out the material. This whould also solve the problem that accure when manual loading the slots. For exemple if you have the shovel recipt acitve, and put one stack of stone and one stack of sticks, instead of just aborting, it automaticly moves half the stack of stick to the last slot.

* When an hopper (or pipe) ask what slots that it can insert items in, maybe I should lie, and maybe just say that it has one empty slot, and when it ask if it could add stone or stick to that slot (in the case of the shovel) I can answer yes on both, on when the item is inserted, I move it to the right slot.

**Comment by ArturilloD** [comment]
I was also doing tests and it would be good if I could somehow see what I
did so that I could see better what I write to him, I will try to record it
and send it.

What he says about the pipes filling very fast if it is true, the point is
that the pipes can simply be told to send a specific amount per item, for
example a stack of each type of item or to send only four of each one ,
even one stone and two sticks at a time, the disadvantage of this is when
you want to do it with several factories at the same time or hoppers, since
the items are moving and it is not possible to distribute it in equal parts.

So that you can see what I'm saying that it would be good if only the
factory accepted one item per slot, you should try to do the tests that you
have done but with the speedy hoppers hopper, at least with the diamond one
it works wonderfully, really, with the others I don't know.

**Comment by puggan** [comment]
v1.2.1 Connected side now ignored, can excrat items at the top, and insert item in the bottom. 
![2020-08-29_16 24 18](https://user-images.githubusercontent.com/866668/91639066-2d641280-ea14-11ea-843b-628a16c30146.png)

**Comment by puggan** [comment]
v1.3.0 implement the auto-rebalance, but now there is a new bug with the pipelines, items get lost :-(.

**Comment by ArturilloD** [comment]
If you mean that they stay inside the table that happened before too, which
will not be reproduced

**Comment by puggan** [comment]
Built a Loop with dimonds and dimond blocks, but in 9 stack of dimond blocks (9x64=576).
After AFK-ing for a bit, I stoped the loop by disabeling one extractor, and now there is less then 1½ stack left (92 blocks).
![2020-08-29_22 47 20](https://user-images.githubusercontent.com/866668/91646894-8bb0e580-ea54-11ea-919a-5637a4a595b7.png)

**Comment by ArturilloD** [comment]
I'm trying to recreate but the chest items won't come out

**Comment by puggan** [comment]
That was a bug on my side.

1. The pipes didn't connect
2. An uneven fill, can make it produce less items, as there can be enough materials, but in the wrong slots.
3. For extracting items, it would be nice if the direction is ignored.

I think all 3 issues above, now is solved in v.1.3.1, Please test it:
https://www.curseforge.com/minecraft/mc-mods/workbench-factory/files/3045386

**Comment by ArturilloD** [comment]
Yes, I also think that being able to extract from any direction is the most
appropriate, since it allows more play. Right now I am testing it, I would
like to be able to test it with the xnet mod, but it has many problems and
I don't know what to do, it seems that they don't pay much attention to it
...

**Comment by ArturilloD** [comment]
Now I will send you a report in a while, I am doing the tests, I would like
to know if there is any way to speed up the process of distributing the
items on the grid, thank you.

**Comment by ArturilloD** [comment]
I did some tests and with xnet they lose a lot, it seems that even if there
is no
space in the factory inventory they still enter? I don't know very well,
then I do more tests, apparently with pretty pipea it ends up returning the
same diamonds, heh. Then I give a lot to pretty, we'll see how they do it

>

**Comment by ArturilloD** [comment]
I don't know yet where the error is, the question is that items disappear.

I can't find out when that happens, from what I have managed to discover,
it seems that the pipes try to insert items and there is no space in the
factory, then they jump, I don't understand very well because sometimes
they return to the original chest and others insert it anyway, So I don't
know if it's from the factory or from the pipes.

Then there is also another error that I have not been able to reproduce and
that is that the items do not enter the factory until the GUI is opened,
although it is already configured.

I am sending you some clips and an image with the mods that I have
installed to do the tests, it is to make sure that it cannot be due to
incompatibility, I have tried to test it with the XNet mod, although with
that there are always items missing, so I suspect that it may be that, I
still have to test more and with other pipes.

The clip on the right side of the GUI is to make a suggestion / request and
is if it would be possible to activate it so that when the table is
deactivated it can be used as a normal work table, please.

PS: Also if you try to use the cookbook the recipe is generated on the
table but without the output element and I have not tried with JEI,
although I think it would be nice to be able to do something with that too,
as if it were the cookbook , select a recipe and it will be automatically
generated on the table.

I will send you what I discover, thanks in advance.
 Base Profile 2020.08.31 - 08.14.04.01.mp4
<https://drive.google.com/file/d/15v5j-elTnnV1xMS_XqDNq6Dug9DWuRK-/view?usp=drive_web>
 Base Profile 2020.08.31 - 08.29.11.04.mp4
<https://drive.google.com/file/d/1KSGmTjb7H3Wr6IrxQCeWdv2cTLHoKRY7/view?usp=drive_web>
 Base Profile 2020.08.31 - 08.30.17.05.mp4
<https://drive.google.com/file/d/16ksPC5RvXbXvspIyzentOBKfhA1wCWWO/view?usp=drive_web>
 Base Profile 2020.08.31 - 08.41.02.06.mp4
<https://drive.google.com/file/d/1_KgathC3qIqSFV-Z6S4tCIde4Vkuyud4/view?usp=drive_web>
 Base Profile Screenshot 2020.08.31 - 08.23.50.8...
<https://drive.google.com/file/d/1qIMJ-e2b9SKeZJ75hGHPe-u-dYEdMeiu/view?usp=drive_web>

**Comment by ArturilloD** [comment]
In these two clips it is seen as in a factory the items do not enter until
I open the GUI and in the other one you see a moment in which blocks jump
and leave the system.
 Base Profile 2020.08.31 - 08.47.03.07_Trim (2).mp4
<https://drive.google.com/file/d/1NqcwNJ32sg9uHdR4XT7mgJMcfkftaiTl/view?usp=drive_web>

 Base Profile 2020.08.31 - 08.47.03.07_Trim.mp4
<https://drive.google.com/file/d/1w9Oy-5ELI_K5ibaHulWQ60B_gxvQz7rb/view?usp=drive_web>

**Comment by ArturilloD** [comment]
Already have access?

**Comment by ArturilloD** [comment]
I have managed to discover the reason why items are lost when crafting with
the Factory and PrettyPipes, the question is that the items come out and go
to the next block, in this case another Factory, as the Factory does not
have space to more items then the Pipes what they do is return the items to
the block they came from, that's when the Pipes cannot insert items, since
the input slots have a different item from the output slot and the output
slot does not allow enter items. Even if items were allowed to enter the
output slot, the problem would still remain somewhat since it cannot fit
more than one stack.

Demo video: https://youtu.be/XQ51T-voYAY

Regarding the disappearance of elements with XNet (
https://www.curseforge.com/minecraft/mc-mods/xnet) I do not know very well
what happens when crafting, what I suppose happens is that they do not
enter more elements and then in that process the excess elements are
eliminated, I still have to do tests, although it is quite exasperating.

**Comment by ArturilloD** [comment]
When I referred to there being only one item per box, I was referring to
this, although I don't quite understand why it doesn't have the same
behavior as the other table. (https://youtu.be/DA4ewRKhots)
